src/reliable_text.c

```diff
@@ -0,0 +1,444 @@
+    float incomingData[LDPC_TOTAL_SIZE_BITS];
+    float llr[LDPC_TOTAL_SIZE_BITS];
+    unsigned char output[LDPC_TOTAL_SIZE_BITS];
+    int parityCheckCount = 0;
+    
+    if (obj->bit_index == obj->sym_index * 2)
+    {
+        // Use soft decision for the LDPC decoder.
+        float EsNo = 3.0; // note: constant from freedv_700.c
+        
+        symbols_to_llrs(llr, (COMP*)&obj->inbound_pending_syms[0], &obj->inbound_pending_amps[0],
+                        EsNo, obj->fdv->ofdm->mean_amp, LDPC_TOTAL_SIZE_BITS / 2);
+    }
+    else
+    {
+        // We don't have symbol data (likely due to incorrect mode), so we fall back
+        // to hard decision.
+        for (int bitIndex = 0; bitIndex < LDPC_TOTAL_SIZE_BITS; bitIndex++)
+        {
+            // Map to value expected by sd_to_llr()
+            incomingData[bitIndex] = 1.0 - 2.0 * src[bitIndex];
+        }
+    
+        sd_to_llr(llr, incomingData, LDPC_TOTAL_SIZE_BITS);
+    }
+    run_ldpc_decoder(&obj->ldpc, output, llr, &parityCheckCount);
+    
+    // Data is valid if BER < 0.2
+    float ber_est = (float)(obj->ldpc.NumberParityBits - parityCheckCount)/obj->ldpc.NumberParityBits;
+    int result = (ber_est < 0.2);
+        
+    //fprintf(stderr, "BER est: %f\n", ber_est);
+    if (result)
+    {        
+        memset(dest, 0, RELIABLE_TEXT_BYTES_PER_ENCODED_SEGMENT);
+        
+        for (int bitIndex = 0; bitIndex < 8; bitIndex++)
+        {
+            if (output[bitIndex])
+                dest[0] |= 1 << bitIndex;
+        }
+        for (int bitIndex = 8; bitIndex < (LDPC_TOTAL_SIZE_BITS / 2); bitIndex++)
+        {
+            int bitsSinceCrc = bitIndex - 8;
+            if (output[bitIndex])
+                dest[1 + (bitsSinceCrc / 6)] |= (1 << (bitsSinceCrc % 6));
+        }
+    }
+    
+    return result;
+}
+
+static void reliable_text_freedv_callback_rx_sym(void *state, _Complex float sym, float amp)
+{
+    reliable_text_impl_t* obj = (reliable_text_impl_t*)state;
+    
+    // Ignore symbol if we're still trying to lock onto UW.
+    if (!obj->uw_bits_valid) return;
+    
+    // Save the symbol otherwise. We'll use it during the bit handling below.
+    obj->inbound_pending_syms[obj->sym_index] = (complex float)sym;
+    obj->inbound_pending_amps[obj->sym_index++] = amp;
+    
+    //fprintf(stderr, "Got sym: %f, amp: %f\n", sym, amp);
+}
+
+static void check_uw(reliable_text_impl_t* obj)
+{
+    // Count number of errors in UW.
+    unsigned short xor_uw = obj->uw_value ^ 0xFFFF;
+    int num_zeroes = 0;
+    for (int bit = 0; bit < RELIABLE_TEXT_UW_LENGTH_BITS; bit++)
+    {
+        if (xor_uw & 0x8000)
+        {
+            num_zeroes++;
+        }
+        xor_uw <<= 1;
+    }
+    obj->uw_bits_valid = obj->bit_index >= RELIABLE_TEXT_UW_LENGTH_BITS && num_zeroes <= RELIABLE_TEXT_MAX_ZEROES_IN_UW;
+}
+
+static void reliable_text_freedv_callback_rx(void *state, char chr)
+{
+    //fprintf(stderr, "received char: %d\n", (chr & 0x3F));
+    
```

Comment:
Perhaps consider a state/next_state switch/case state machine implementation in this function for the acquisition (looking for UW) and synced states.  Tends to work out neater.  Many examples in other parts of codec2.

Reply:
I ended up redoing how UW gets sync'd anyway due to issues I found while trying to implement interleaving. Will push it up shortly.